baselines/jft/al_utils.py

```diff
@@ -12,81 +12,134 @@
         record['id'] = tf.constant(int_id)
         yield record
 
   return inner
 
 
-class Cifar10Subset(cifar.Cifar10):
-  """CIFAR-10 Subset.
-
-    Implement CIFAR-10 with the added functionality taking a subst by id.
+class SubsetDatasetBuilder(DatasetBuilder):
+  """Subset Dataset Builder which is "just right" for clu.
 
     Args:
+      base_dataset_builder: a DatasetBuilder for the underlying dataset (essentially an object with a as_dataset method)
       subset_ids: a dictionary of split: set of ids.
-
-    Returns:
-      A Cifar10Subset object.
   """
 
-  # TODO(trandustin, wangzi): Add type annotation for subset_id that works.
-  # def __init__(self, *, subset_ids: Dict[str, Set[int]], **kwargs):
-  def __init__(self, *, subset_ids, data_dir=None, **kwargs):
-    super().__init__(data_dir=data_dir, **kwargs)
-    self.subset_ids = subset_ids
+  def __init__(self, base_dataset_builder: tfds.core.DatasetBuilder, *,
+               subset_ids: Optional[Iterable[int]]):
+    self.subset_ids = set(subset_ids) if subset_ids is not None else None
+    self.base_dataset_builder = base_dataset_builder
+
+  def as_dataset(self,
+                 split: Union[str, tfds.core.ReadInstruction],
+                 *,
+                 shuffle_files: bool = False,
+                 read_config: Optional[tfds.ReadConfig] = None,
+                 **kwargs) -> tf.data.Dataset:
+    # We don't allow an empty split by virtue of the parameter declaration, so we always have a split.
+    read_config = dataclasses.replace(
+        kwargs.pop('read_config', tfds.ReadConfig()))
+    # Add the 'tfds_id' key to the samples which we can then parse.
+    read_config.add_tfds_id = True
+
+    dataset = self.base_dataset_builder.as_dataset(
+        split=split, shuffle_files=False, read_config=read_config, **kwargs)
 
-  def _as_dataset(self, *args, split, **kwargs):
-    dataset = super()._as_dataset(*args, split=split, **kwargs)
-
-    # HACK: Fix id to be an int. Assuming 'label' is an int, too.
     element_spec = dataset.element_spec.copy()
+    element_spec['id'] = tf.TensorSpec(shape=(), dtype=tf.int64, name=None)
     logging.info(msg=f'element_spec = {element_spec}; '
                  f'type = {jax.tree_map(type, element_spec)}')
-    element_spec['id'] = element_spec['label']
-
-    # NOTE: if this line errors out, make sure to update your
-    # tensorflow-datasets package to the right version.
-    split_name = split.split
 
-    return tf.data.Dataset.from_generator(
-        _subset_generator(dataset, self.subset_ids[split_name]),
+    dataset = tf.data.Dataset.from_generator(
+        _subset_generator(
+            dataset=dataset,
+            dataset_info=self.base_dataset_builder.info,
+            subset_ids=self.subset_ids,
+            splitwise_id=True),
         output_signature=element_spec,
-    ).cache()
+    )
+
+    # This is a bit more complex: potentially cache before or after calling .shuffle.
```

Comment:
I don't follow the below, did you base it on some other code? Would be good to reference that here.

Reply:
It's handwritten and follows the logic of reshuffling on every iteration or not. If we call cache before the shuffle, each epoch shuffle will be executed. Otherwise, if we call cache after shuffle, it will fix the first shuffle in place (in line with the behavior of that readconfig).